Refactor `HTTP::Server` to bind to multiple addresses

straight-shoota opened the pull request:
Until now, a `HTTP::Server` instance binds to exactly one IP address (host + port). This address was set in the constructor alongside several options for passing one or more handlers.

This resulted in quite a few constructors already, and adding support for more options (such as using a Unix socket) would multiply the numbers. Additionally, it should be possible to have the server listen on multiple addresses, as many other HTTP server implementations.

This PR changes the API of `HTTP::Server`:

* The constructor only accepts handlers (in multiple variations, as before).
* `#bind` is used to add one ore multiple addresses the server should listen on. It accepts the same arguments previously used in the constructor (`port`, `host, port` - each with optional `reuse_port`).
* Additional overloads exist for `Socket::IPAddress` and `Socker::Server`.
* <del>You can bind to a Unix socket using `#bind(Socket::UNIXAddress)` or by supplying a `UNIXServer`</del> (<ins>will be a subsequent PR</ins>)
* <del>There is also a single-argument overload that accepts a string with either a host + port or a Unix socket (prefixed by `unix:`) and creates the appropriate server accordingly. See [#2735 (issue-comment)](https://github.com/crystal-lang/crystal/issues/2735#issuecomment-301349490) for the reasoning behind this.</del> (<ins>will be a subsequent PR</ins>)
* `#bind_unused_port` is a shortcut for `bind(0)` and returns the port. This is particularly useful for the use case of binding to an unused port and printing the port (could maybe change the return type to `Socket::IPAddress` to include the address as well). `#port` was removed because there is no single port anymore that this method could return.
* `#listen` starts accepting incoming requests, as before. However, there is also an catch-all overload which forwards all arguments to `#bind` and starts listening. This is a shortcut for the following:
```crystal
server.bind 80
server.listen
# same as
server.listen 80
```

With a server listening on multiple interfaces, I figured it would be useful to be able to determine from which connection a client request originates, so I added a nilable (for stubbing) instance variable `socket` to `HTTP::Server::Context`.

<del>This also includes a minor fix to copy the path of a UNIXServer to each socket. Otherwise you couldn't really tell them apart if (only by the file descriptor).</del> (<ins>will be a subsequent PR</ins>)

<del>Fixes one half of #2735</del> (<ins>will be a subsequent PR</ins>)

RX14 commented on src/http/server.cr:
```diff
@@ -95,72 +100,187 @@ class HTTP::Server
 
-  def self.new(port, &handler : Context ->)
-    new("127.0.0.1", port, &handler)
+  # DEPRECATED: Create an instance and call `#bind` instead.
```
We don't deprecate before 1.0, we just remove. Unless it's super duper common and used in many places in one project (think `MemoryIO`), and even then we make the overloads print notices using macros otherwise nobody will notice.

RX14 commented on src/http/server.cr:
```diff
@@ -95,72 +100,187 @@ class HTTP::Server
+  def bind(address : String, reuse_port : Bool = false) : Socket::Server
+    if address.starts_with?("unix:")
+      {% if flag?(:unix) %}
+      bind(Socket::UNIXAddress.new(address[5..-1])).as(Socket::Server)
```
I'm not sure we want to continue spattering ifdefs around the codebase. If sockets aren't supported, we'll raise in `UNIXAddress.new` instead of everywhere it's used. Indent is wrong.

straight-shoota replied:
Yeah, it's probably better this way. The whole Socket API looks like it needs a major refactoring anyway.

ysbaddaden replied:
Some duplication, here. Having different `#bind` overloads is already redundant, but introducing yet another one with a URI-like string, required for binding UNIX sockets, that feels too much and weirdly designed.

It should either be the only solution, or there should be explicitly named methods (ala puma), which IMHO feels better (an application may require a URI, parse it, and call the correct methods):

- `#bind_tcp_listener(host, port)`
- `#bind_ssl_listener(host, port, ctx)`
- `#bind_unix_listener(host, port)`

straight-shoota replied:
The UNIX socket part is removed from this PR. Regarding method naming, I'd suggest to continue the discussion in the main thread.

RX14 commented on src/http/server.cr:
```diff
@@ -95,72 +100,187 @@ class HTTP::Server
+  # Creates a `Socket::Server` listenting to *address* and adds it as an interface.
+  #
+  # The *address* can either be `host:port` or a Unix socket prefixed by `unix:`.
+  # This obviously works only on systems supporting Unix sockets.
```
"obviously" is unneccesary.

RX14 commented on src/http/server.cr:
```diff
@@ -95,72 +100,187 @@ class HTTP::Server
   end
 
-  # Creates the underlying `TCPServer` if the doesn't already exist.
+  {% if flag?(:unix) %}
```
Don't bother.

RX14 commented on src/http/server.cr:
```diff
@@ -95,72 +100,187 @@ class HTTP::Server
     until @wants_close
-      spawn handle_client(server.accept?)
+      @interfaces.each do |interface|
+        spawn handle_client(interface.accept?)
```
This is completely wrong. It will round-robin each interface accepting one connection from each interface in order, instead of which interface is ready to accept. You need to spawn a fiber for each interface.

straight-shoota replied:
Darn. I'll put my shame hat on and hide in the basement... :(
The worst of it is that I had a relevant spec for this which somehow got lost during rebasing.

RX14 commented on src/http/server/context.cr:
```diff
@@ -7,8 +7,11 @@ class HTTP::Server
     getter response : Response
 
+    # An optional `Socket` which holds the client connection of this context.
+    getter socket : Socket?
```
I'm not sure about this.

ysbaddaden replied:
Same here. Hijacking the IO must be achieved within a request handler or response output (see WebSocket handler).

straight-shoota replied:
The websocket handler really hijacks the IO using `upgrade`. The problem I'm trying to solve is not about taking over the IO handling (including responsibility for calling `#close`). It's rather to get information about the connection, such as `local_address`. Don't know about other things... `remote_address` could be useful.
So maybe we could just expose those in `Response`?

RX14 replied:
I'd prefer to expose certain metadata regarding the socket, instead of an actual socket you can perform IO on.

RX14 commented:
Also I strongly recommend doing the UNIX socket stuff (new feature) in a later PR and leaving this just as a pure refactor.

straight-shoota commented:
Sure, I can split it up.

I'm not entirely sure about `Context#socket` but I'm not sure there is an alternative and I think the socket should be accessible from a handler when you can have different ones.

ysbaddaden commented on src/http/server.cr:
```diff
@@ -95,72 +100,187 @@ class HTTP::Server
+  # You may set *reuse_port* to true to enable the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind(port : Int32, reuse_port : Bool = false) : TCPServer
+    bind "127.0.0.1", port, reuse_port
```
I was gonna say that binding to `127.0.0.1` makes the `#bind(port)` overload somewhat useless (invisible on network, doesn't bind ipv6 interfaces) but that's how it's currently done...

That's still weird, thought, since `TCPServer.new(port)` will bind on `::` (all ipv4 & ipv6 interfaces) for example.

straight-shoota replied:
Yes, this should probably be improved. Although I'm not entirely sure if `::` would be a sane default.
I'd suggest to keep the current behavior for now and open an issue about this to fix it once this is merged.

RX14 replied:
I prefer to keep everything local-addressed by default, and force people to choose possibly insecure defaults (0.0.0.0, ::) by conscious decision.

straight-shoota replied:
Still, we should make sure to take the IPv6 loopback `::1` into account as well. It would probably nice to have a constant or class method in `Socket::IPAddress` instead of using magic value here.

straight-shoota replied:
Maybe we could have a method to listen on `0.0.0.0` and `::`, like `#bind_all(port : Int32)`

ysbaddaden commented on src/http/server.cr:
````diff
@@ -95,72 +100,187 @@ class HTTP::Server
+  # server = HTTP::Server.new { }
+  # server.bind Socket::UNIXAddress.new("/tmp/my-socket")
+  # ```
+  def bind(address : Socket::UNIXAddress) : UNIXServer
````
Are the XAddress overloads *really* useful?

straight-shoota replied:
I guess we don't necessarily need them, it's just that the Socket API is not really nice in that regard, you currently can't directly create a UNIXSocket from a UNIXAddress. But this needs to be fixed there (see #5778) then we can probably do without the overloads. Or have only a general `bind(address : Socket::Address)` which delegates creating the appropriate server to `Socket::Server.new(address : Socket::Address)`. That would probably be the best solution.

ysbaddaden commented on src/http/server.cr:
```diff
@@ -95,72 +100,187 @@ class HTTP::Server
+  end
+
+  # Adds a `Socket::Server` *interface* to this server.
+  def bind(interface)
```
There should be a `#bind(interface : Socket::Server)` type constraint.

ysbaddaden commented on src/http/server.cr:
```diff
@@ -95,72 +100,187 @@ class HTTP::Server
+
+  # Binds to the specified arguments and starts the server.
+  #
+  # See `#bind` and `#listen` for details
```
Is it really helpful? We're introducing and require explicit `#bind` but in the same time we mix its logic (bind an interface) with `#listen` (for incoming connections).

straight-shoota replied:
It's somewhat nice but not really necessary, I think. I guess we don't need it and it convolutes the API.

I will remove it. I desired, it can be added later.

ysbaddaden commented on src/http/server.cr:
```diff
@@ -169,10 +289,12 @@ class HTTP::Server
-      @server = nil
+
+    @interfaces.each do |interface|
+      interface.close
```
Closing an interface can raise, which will prevent further interfaces from being closed.

straight-shoota replied:
Will wrap it in rescue. We can probably ignore any exceptions, right?

straight-shoota commented:
@ysbaddaden Unfortunately, I have pushed new commits while you were reviewing and your comments have been hidden but not all of them are resolved.

The `#bind` overloads are reduced to a minimum. I don't like having different method names for this because it's just not necessary when the arguments clearly convey the meaning. I don't see much sense in using different names for the same thing.

In your examples, there was also a mention of `bind_tls_socket` - do you think we need tls based on individual sockets? I'd favour having the entire server run either with TLS or not.

RX14 commented:
I like the `listen` overload that forwards to `bind`. The difference between
```cr
server = HTTP::Server.new do |ctx|
  ctx.response.puts "Hello World!"
end

server.bind(80)
server.listen
```

and

```cr
server = HTTP::Server.new do |ctx|
  ctx.response.puts "Hello World!"
end

server.listen(80)
```

is actually quite large, as learning an ordering of methods - rather than simply a single method - is a fair bit more of a hurdle.

straight-shoota commented:
It definitely has something to it. To me, the strongest point I see against it is the method signature. It would either delegate every single overload of `#bind` - which means a lot of overloads for `#listen` as well - or as I had it implemented, just catch the arguments and forward them to `#bind`. This is easier but `#listen`'s method signature does no longer show what arguments it can accept.
But this might not be such a huge issue after all... hm. The docs can just list all possible overloads and/or reference `#bind` signatures.
In the end, it is a nice feature for about 90% of users who just need to listen on one socket.

jhass commented:
Not sure I'm entirely convinced of it myself, but just as another possibility, we could also have `bind`, `listen` without arguments and a `bind_and_listen` that forwards its argument to `bind` for the common case

straight-shoota commented:
Hm, that would make the name for the common case very bulky...

straight-shoota commented:
I've removed the commit exposing `socket` on `HTTP::Server::Context` and will open a new issue for that.

jhass commented:
Maybe we can find a less bulky but not too confusing name? `start` or `serve` would come to mind.

straight-shoota commented:
I doubt it would be helpful to introduce a different name. The main purpose is to `listen` - it just happens to also add a socket before that. We shouldn't have two method names for listening (which is also blocking).

straight-shoota commented:
What do you think about the return value of `#bind_unused_port`? It would make sense to return an `IPAddress` (ip + port) instead of only a port. That's especially true when called without a `host` argument, you shouldn't have to guess which IP is used. And the typical usecase will be printing the address (ip + port) like this:
```crystal
address = server.bind_unused_port
puts "Listening on http://#{address}/"
server.listen
```

Even if the host is specified, it still makes sense to return the entire address (instead of having different return values depending on the presence of `host` argument).

And I guess it would be great to have access to the `sockets` ivar, or maybe only the local addresses of the sockets. If the server is listening on multiple sockets, the equivalent to the above sample could be:
```crystal
# bind to a few sockets (maybe from config settings)

server.local_addresses.each do |address|
  puts "Listening on http://#{address}/"
end
server.listen
```

straight-shoota commented:
I've changed the return type of all `bind` and `bind_unused_port` method to `Socket::IPAddress`. That's what's typically interesting (to get the actual adress if port is zero) and there doesn't need to be direct access to the `TCPServer`. When this should really by necessary, it is possible to bind a custom `TCPServer` instance.

straight-shoota commented:
I'm not sure if `bind(server : TCPServer)` should maybe be renamed to `add(server : TCPServer)` or `<<(server : TCPServer)`.

ysbaddaden commented:
Starting individual TCP and SSL servers is half the point of having multiple listeners to me. Without it, your patch would be pointless in Prax for example (binds a single HTTP::Server to both HTTP and HTTPS).

ysbaddaden commented:
Here is a confusing example:

```crystal
server.bind("localhost")
server.bind("/tmp/app.sock")
```

straight-shoota commented:
I'd wan't to keep it a simple as possible. Keeping track of which sockets need to use TLS inside `HTTP::Server` seems unnecessarily tedious to me.

What about if we just use a special SSL server socket for this? It would essentially wrap another `Socket::Server` (`TCPServer`) and return a `OpenSSL::SSL::Socket::Server` wrapping a `Socket` instead of the `TCPSocket` directly.
This would make the use of TLS completely transparent from the perspective of the `HTTP::Server` and would be reusable as a drop-in replacement for a plain TCPServer (of course there should be a `bind` overload for setting this up which receives a SSL context).


@RX14 asked to introduce Unix sockets in a subsequent PR and I agree. This let's us focus on the general API first (obviously having upcoming Unix sockets in mind) and worry about Unix socket specifics separately.


Your confusing example doesn't work because you can't bind to a host without a port. It would be:
```crystal
server.bind("localhost", 80)
server.bind("/tmp/app.sock")
```

The only overlap would be an overload accepting a String as generic URI like `tcp://localhost:80`, `unix:///tmp/app.sock` or `ssl://0.0.0.0:442` vs. an overload that expects a String as path for a Unix socket.
So, yeah maybe this calls for differently named methods. The downside of this is a more complicated API (a few different method names instead of semantically matching overloads) and it wouldn't be as easy to forward calls from `listen` which equally helps to provide a simple API as per [@RX14's comment](https://github.com/crystal-lang/crystal/pull/5776#issuecomment-370981929).

I'd probably rather think about having a different overload for generic URI strings, only accepting `URI` type without a named argument or delegate this entirely to `Socket::Address`:
```crystal
server.bind Socket::Address.parse "tcp://127.0.0.1:80"
server.bind Socket::Address.parse "unix:///var/run/socket"
```

You're right, `#bind_unused_port` doesn't add much over specifying `port: 0` anymore. I still like it for it's explicit meaning instead of relying on a magic number.

ysbaddaden commented:
**Design**:
We *must* decide how to integrate UNIX sockets. They're part of the reason for binding multiple sockets. Same for HTTPS, it makes no sense to me to force the same TLS context for all bound sockets —bummer, I won't be able to leverage this patch! Resolving these issues now makes more sense (to me) than delaying to some further pull request. Let's break things once and thoroughly think the whole problem.

**Confusing example**:
I specifically said that "you must specify a port" isn't an acceptable answer. If we add `#bind(path : String)` for UNIX sockets then `bind("localhost")` will silently bind an UNIX socket. That's confusing and totally silent: no error is ever raised, despite not working as the developer intended.

A `#bind(uri : String)` can be confusing too, unless we force `uri.scheme` to be defined, so `bind("localhost")` will raise an "invalid socket URI" runtime error, and I can lookup the documentation to understand the issue.

We can introduce specifically named bind methods, a generic `#bind(URI)`, and still introduce `#listen(port: Int32)` for the most general use case: bind and listen on a single port on all interfaces. We don't have to simplify for all use cases, just the most common one; or we don't and consider that the whole benefit is worth the extra line.

Having a `SSLServer` (or something like that) class that would instantiate/wrap a `TCPServer` then wrap the accepted `TCPSocket` is a great idea. `HTTP::Server` as well as any other servers shouldn't have to deal with that.

I'm sure we need to deal with a `Socket::Address.parse` method. The following seems far enough to me:
```crystal
uri = URI.parse(str);
case uri.scheme
when "tcp"
  host = uri.host == "" ? "::" uri.host
  add TCPServer.new(host, uri.port)
when "unix"
  add UNIXServer.new(uri.path)
else
  raise ArgumentError.new("invalid socket URI")
end
```

ysbaddaden commented:
wrong close, sorry

straight-shoota commented:
I'm okay with re-adding Unix sockets to this PR and I think we agree on `SSLServer`/`SSLSocket` and `Socket::Address.parse`. About the `bind` methods I'm not decided, but I see your point and I think we could probably go that way. Let's hear what others think about it ;)

I like the idea to generalize the concept of a server for multiple sockets. I'm not sure we actually need an additional type for that. In fact, the only thing specific to HTTP in `HTTP::Server` are the constructors (which create a `RequestProcessor` from handlers). You could simple create a subclass that sets `@processor` to a custom request processor implementation and serve a different protocol.

So maybe it would be an option to rename `HTTP::Server` to `MultiServer` (or similar; I don't know why it would need generics) and move `HTTP::Server::RequestProcessor` (which handles the HTTP-protocol) to `HTTP::Server`. The request processor would become the typical API to create an HTTP server. It would wrap a `MultiServer` and delegate `bind` and `listen` methods.

straight-shoota commented:
@RX14 @jhass WDYT?

RX14 commented on src/http/server.cr:
```diff
@@ -95,90 +100,107 @@ class HTTP::Server
+  # Creates a `TCPServer` and adds it as a socket, returning the local address
+  # and port the server listens on.
+  #
+  # If *port* is `0`, a random, free port will be chosen.
```
If you're going to have `bind_unused_port` then don't advertise this. Consider even disallowing it. It's platform-specific.

RX14 commented on src/http/server.cr:
```diff
@@ -95,90 +100,107 @@ class HTTP::Server
-    server = bind(reuse_port)
-    until @wants_close
-      spawn handle_client(server.accept?)
+  def listen
```
Should forward it's args to bind.

RX14 commented:
This PR doesn't need to support SSL or unix sockets. They just need to be designed and ready to go.

straight-shoota commented:
I have updated this PR and renamed `#bind` methods receiving port and host+port to `#bind_tcp` and added overloads to `#listen` with these signatures.

The following still needs to be done (in separate PR's):
* Add `#bind_unix` to bind to a `UNIXServer`
* Implement `SSLServer` to add an SSL context per socket (wrapping TCP / UNIX server)
* Implement `Socket::Address.parse` and add `#bind(uri : URI | String)`
* Extract generalized `MultiServer` interface

sdogruyol commented:
I think this should be ready for another round of review :+1:

ysbaddaden commented on src/http/server.cr:
```diff
@@ -95,90 +100,140 @@ class HTTP::Server
+  end
+
+  # Adds a `Socket::Server` *socket* to this server.
+  def bind(socket : Socket::Server) : Socket::Server
```
Why return the server?

straight-shoota replied:
It's intended as a simplification. 
This makes it easier to retrieve the actual address from binding a server whose local address depends on binding it: `http_server.bind(TCPServer.new(0)).address`

straight-shoota replied:
But yeah, it's not such a clean API and can probably be removed.

RX14 replied:
I vote remove it, variables work fine.

ysbaddaden commented on src/http/server.cr:
```diff
@@ -95,90 +100,140 @@ class HTTP::Server
+  #
+  # If *reuse_port* is `true`, it enables the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind_tcp(port : Int32, reuse_port : Bool = false) : Socket::IPAddress
```
Why return the address?

straight-shoota replied:
To get the actual address including the automatically assigned port when `port == 0`.

straight-shoota replied:
While there is also `bind_unused_port`, it can't be used if `port` is configurable (at least not directly).

ysbaddaden commented on src/http/server.cr:
```diff
@@ -95,90 +100,140 @@ class HTTP::Server
+      end
     end
+
+    @sockets.size.times { done.receive }
```
What if `#accept?` raises? It can raise an `Errno` for example.

straight-shoota replied:
Should be fixed now.

RX14 commented on src/http/server.cr:
```diff
@@ -95,90 +100,140 @@ class HTTP::Server
+          spawn handle_client(socket.accept? || break)
+        end
+
+        done.send nil
```
put this in `ensure`?

straight-shoota commented:
I'm happy to add Unix and SSL sockets as well if favoured.

straight-shoota commented on spec/std/http/server/server_spec.cr:
```diff
@@ -219,15 +217,47 @@ module HTTP
+
+      HTTP::Client.get("http://#{address2}/").body.should eq "Test Server (#{address2})"
+      HTTP::Client.get("http://#{address1}/").body.should eq "Test Server (#{address1})"
+      HTTP::Client.get("http://#{address1}/").body.should eq "Test Server (#{address1})"
```
It's actually supposed to ensure that it doesn't simply alternate between addresses or something strange...

bcardiff commented on src/http/server.cr:
```diff
@@ -95,90 +100,138 @@ class HTTP::Server
+
+  # Starts the server. Blocks until the server is closed.
+  def listen
+    raise "Can't start server with not sockets to listen to" if @sockets.empty?
```
Maybe we should include something like "Use HTTP::Server#bind" in the error so we can guide the user a bit. WDYT?

straight-shoota replied:
Another issue, which should probably also raise, is when `bind` is called after `listen`:
```cr
server.bind socket
spawn { server.listen }
server.bind socket2
```

For this we'd need a ivar to track if the server is listening. It could probably be implemented to retroactively start listening on the additional socket as well, but that's more complicated and probably not that useful without having service objects.

bcardiff replied:
Yeap, I would say to add an ivar to track that too.

bcardiff commented:
Given that argless `#listen` is still there, wasn't the api of constructing the server with an initial/single host/port and then calling listen comfortable enough? 

I fail to see the reason why to get rid of that constructor.

I think that easier/shorter interfaces for common use case are important. And add more fine grained ones to allow more rare scenarios, not at the expense of losing the easier ones.

straight-shoota commented:
@bcardiff It's comfortable, but such an additional signature for `new` would add four additional overloads. I'd like to avoid that. Using the `listen` short cut should be equally comfortable and easier to transform to `bind` calls:

```cr
server = HTTP::Server.new(host, port) { }
server.listen
# vs.
server = HTTP::Server.new { }
server.listen(host, port)
```

straight-shoota commented:
You'd need to have these four overloads, even if you just want to support `host : String, port : Int32`.

* `self.new(host : String, port : Int32, &handler : HTTP::Handler::Proc)`
* `self.new(host : String, port : Int32, handlers : Array(HTTP::Handler), &handler : HTTP::Handler::Proc)`
* `self.new(host : String, port : Int32, handlers : Array(HTTP::Handler))`
* `initialize(host : String, port : Int32, handler : HTTP::Handler | HTTP::Handler::Proc)`

I don't really see how specifying the host and port in the constructor is a simpler introduction to the http server than in the `#listen` method (see my previous post). I rather think it's the other way around.

bcardiff commented:
I was thinking just in the first overload. I haven't seen the other used that much in presentations and smalls projects. But, ok. Let's keep the proposal (and update website homepage sample on release 😉 )

RX14 commented on src/http/server.cr:
```diff
@@ -159,6 +162,9 @@ class HTTP::Server
   # Adds a `Socket::Server` *socket* to this server.
   def bind(socket : Socket::Server) : Nil
+    raise "Can't add socket to running server." if listening?
+    raise "Can't add socket to closed server." if @wants_close
```
Just disallow `#close` `unless listening?`. I'd also just rename this variable to `@closed` because the state persists after it's fully shut down - not just when it's in some "closing down" state that "wants close" implies.

straight-shoota replied:
You need to be able to call `#close` even if the server is not listening because sockets may already be bound and need to be released.

Sija replied:
having `getter? closed` would be nice too.

straight-shoota replied:
Already in there ;)

straight-shoota commented:
@RX14 The squashed commit message of 01a5126 just joined all individual commit messages together, including the fixups. It's not a big deal but it would be preferable if at least the fixups were removed as they don't add anything to the message.
